src/ppx_repr/lib/meta_deriving.ml

```diff
@@ -0,0 +1,142 @@
+open Ppxlib
+
+module Plugin = struct
+  type t = {
+    name : string;
+    type_name : [ `before | `after ];
+    impl : location -> expression -> expression;
+    intf : location -> core_type -> core_type;
+  }
+
+  let create ?(type_name = `after) ~impl ~intf name =
+    { name; type_name; impl; intf }
+
+  let op_name_of_type_name t n =
+    match (n, t.type_name) with
+    | "t", _ -> t.name
+    | x, `before -> Printf.sprintf "%s_%s" x t.name
+    | x, `after -> Printf.sprintf "%s_%s" t.name x
+
+  let derive_str t ~loc ~type_name ~params ~expr:repr =
+    let (module Ast_builder) = Ast_builder.make loc in
+    let open Ast_builder in
+    let name = op_name_of_type_name t type_name in
+    let expr =
+      let body = t.impl loc repr in
+      ListLabels.fold_right params ~init:body ~f:(fun p acc ->
+          pexp_fun Nolabel None (pvar p) acc)
+    in
+    pstr_value Nonrecursive
+      [ value_binding ~pat:(ppat_var (Located.mk name)) ~expr ]
+
+  let derive_sig t ~loc ~type_name ~params ~ctyp:repr =
+    let (module Ast_builder) = Ast_builder.make loc in
+    let open Ast_builder in
+    let name = op_name_of_type_name t type_name in
+    let type_ =
+      let return_type = t.intf loc repr in
+      ListLabels.fold_right params ~init:return_type ~f:(ptyp_arrow Nolabel)
+    in
+    psig_value (value_description ~name:(Located.mk name) ~type_ ~prim:[])
+
+  let defaults =
+    [
+      create "equal"
+        ~impl:(fun loc t -> [%expr Repr.unstage (Repr.equal [%e t])])
+        ~intf:(fun loc t -> [%type: [%t t] -> [%t t] -> bool]);
+      create "compare"
+        ~impl:(fun loc t -> [%expr Repr.unstage (Repr.compare [%e t])])
+        ~intf:(fun loc t -> [%type: [%t t] -> [%t t] -> int]);
+      create "size_of"
+        ~impl:(fun loc t -> [%expr Repr.unstage (Repr.size_of [%e t])])
+        ~intf:(fun loc t -> [%type: [%t t] -> int option]);
+      create "pp"
+        ~impl:(fun loc t -> [%expr Repr.pp [%e t]])
+        ~intf:(fun loc t -> [%type: Stdlib.Format.formatter -> [%t t] -> unit]);
+      create "pp_dump"
+        ~impl:(fun loc t -> [%expr Repr.pp_dump [%e t]])
+        ~intf:(fun loc t -> [%type: Stdlib.Format.formatter -> [%t t] -> unit]);
+      create "random"
+        ~impl:(fun loc t -> [%expr Repr.unstage (Repr.random [%e t])])
+        ~intf:(fun loc t -> [%type: unit -> [%t t]]);
+      create "to_bin_string" ~type_name:`before
+        ~impl:(fun loc t -> [%expr Repr.unstage (Repr.to_bin_string [%e t])])
+        ~intf:(fun loc t -> [%type: [%t t] -> string]);
+      create "of_bin_string" ~type_name:`before
```

Comment:
Is there a reason why we'd want automatic deriving to prefix or suffix the operation?

Reply:
Mostly just interop with the existing conventions used in Irmin:

```ocaml
type value
val pp_value : ...            (* prefix the operation *)
val value_of_bin_string : ... (* suffix the operation *)
```

These seem sensible to me, so I chose to bake it in here.

`ppx_deriving` has a similar feature in [its conventions](https://github.com/ocaml-ppx/ppx_deriving#plugin-conventions) (although it also provides an option for the user to override the convention, which I think is a bit needless).